packages/synapse-react-client/src/components/QueryWrapper/useOnQueryDataChange.ts

```diff
@@ -0,0 +1,28 @@
+// data is sometimes undefined, which useDeepCompareEffect doesn't like, so use useDeepCompareEffectNoCheck instead
+import { useDeepCompareEffectNoCheck } from 'use-deep-compare-effect'
+import { useQuery, UseQueryOptions } from '@tanstack/react-query'
+import { DefaultError } from '@tanstack/query-core'
+
+type UseOnQueryDataChangeOptions<
+  TQueryFnData = unknown,
+  TError = DefaultError,
+  TData = TQueryFnData,
+> = {
+  useQueryOptions: UseQueryOptions<TQueryFnData, TError, TData>
+  onChange: (data: TData) => void
+}
+
+export default function useOnQueryDataChange<
```

Comment:
New custom hook with logic that was previously in QueryWrapper, extracted to make QueryWrapper as lean as possible.